lxd/devlxd.go

```diff
@@ -118,21 +122,68 @@ var devlxdEventsGet = devLxdHandler{"/1.0/events", func(d *Daemon, c instance.In
 		typeStr = "config,device"
 	}
 
-	conn, err := shared.WebsocketUpgrader.Upgrade(w, r, nil)
-	if err != nil {
-		return &devLxdResponse{"internal server error", http.StatusInternalServerError, "raw"}
+	if r.Header.Get("Upgrade") == "websocket" {
+		conn, err := shared.WebsocketUpgrader.Upgrade(w, r, nil)
+		if err != nil {
+			return &devLxdResponse{"internal server error", http.StatusInternalServerError, "raw"}
+		}
+		defer conn.Close() // This ensures the go routine below is ended when this function ends.
+
+		listener, err := d.devlxdEvents.AddListener(c.ID(), conn, strings.Split(typeStr, ","))
+		if err != nil {
+			return &devLxdResponse{"internal server error", http.StatusInternalServerError, "raw"}
+		}
+
+		logger.Debugf("New container event listener for '%s': %s", project.Instance(c.Project(), c.Name()), listener.ID())
+		listener.Wait(r.Context())
+
+		return &devLxdResponse{"websocket", http.StatusOK, "websocket"}
 	}
-	defer conn.Close() // This ensures the go routine below is ended when this function ends.
 
-	listener, err := d.devlxdEvents.AddListener(c.ID(), conn, strings.Split(typeStr, ","))
-	if err != nil {
+	flusher, ok := w.(http.Flusher)
+	if !ok {
 		return &devLxdResponse{"internal server error", http.StatusInternalServerError, "raw"}
 	}
 
-	logger.Debugf("New container event listener for '%s': %s", project.Instance(c.Project(), c.Name()), listener.ID())
-	listener.Wait(r.Context())
+	w.Header().Set("Connection", "Keep-Alive")
+	w.Header().Set("Content-Type", "application/json")
+
+	// The X-Content-Type-Options response HTTP header is a marker used by the server to indicate
+	// that the MIME types advertised in the Content-Type headers should be followed and not be
+	// changed. The header allows you to avoid MIME type sniffing by saying that the MIME types are
+	// deliberately configured.
+	w.Header().Set("X-Content-Type-Options", "nosniff")
+
+	var lastEvent api.Event
+	messageTypes := strings.Split(typeStr, ",")
+
+	for {
+		event := d.devlxdEvents.LastEvent()
```

Comment:
That feels odd.

I've not done too much digging but can't we add a AddListenerStream or something to the events package and have it handle it almost identically to websocket but with the difference that we're doing http writes in an http stream response?

That would avoid this whole lastEvent thing which feels like a big race condition right now where we're only looking for one event per 500ms and would lose batches of event within a 500ms window (not to mention the wakeups every 500ms when nothing happens).